File: Linked List/92. Reverse Linked List II/92. Reverse Linked List II.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

# 本题的思路较为基础，主要考察链表的反转方法：
# 对于单向链表而言，链表的标准反转方法是依次将原始链表的头结点后的每一个结点插入到原始链表中头结点的前驱结点之前，例如1->2->3, 2->1->3, 3->2->1
# 对于双向链表而言，链表的标准反转方法是直接将每个结点的back指针和next指针的值进行交换

class Solution(object):
    def reverseBetween(self, head, m, n):
        """
        :type head: ListNode
        :type m: int
        :type n: int
        :rtype: ListNode
        """
        # 0. 处理特殊情况：当m==n时，无需反转链表，直接进行返回即可
        if m==n:
            return head

        # 1. 将原始链表增加一个哑结点
        newhead=ListNode(0)
        newhead.next=head
        head=newhead

        pos=head     # 原始链表游标指针

        # 2. 查找待反转部分链表的第一个结点的前驱结点pos
        for i in range(m-1):
            pos=pos.next
        
        # 现在pos指向待反转部分的第一个结点的前驱结点
        now=pos.next  # 创建第二个游标指针now，初始化为待反转链表部分的头节点，注意到m==n的特殊情况已经处理过了，因此pos.next必定不为NULL
        swap=now.next    # pos，now以及swap之间的关系为:pos->now->swap->...，其中now为要反转链表片段的第一个结点

        for i in range(n-m):    # 注意n>=m的基本关系，这里容易出错
            # 将swap插入到待反转部分链表头节点的前驱结点pos之后，pos->now->swap->...反转为pos->swap->now->...
            now.next=swap.next
            swap.next=pos.next
            pos.next=swap

            # 更新下一个要反转插入的结点
            swap=now.next

        return head.next


 

        
        
        
        

File: Linked List/92. Reverse Linked List II/92. Reverse Linked List II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int m, int n) {
        // 1. 特殊处理m==n的情况，这时需要反转的链表片段为空链表
        if(m==n)
        return head;

        // 2. 为原始链表加上哑结点，因为头结点可能也会被反转/修改
        ListNode *newhead=new ListNode(0);
        newhead->next=head;
        head=newhead;

        // 3. 找出待反转部分链表头节点的前驱结点pos，例如对于链表1->2->3->4->5,m=2,n=4，待反转的部分链表为2->3->4，待反转部分链表的头结点的前驱结点为pos=1
        ListNode *pos=head;
        for(int i=0;i<m-1;i++)
        {
            pos=pos->next;
        }

        // 4. 创建第二个游标指针now，从待反转的部分链表的第二个结点开始遍历直到待反转的部分链表的最后一个结点，对于所访问的每个结点，从所在的位置删除该结点，并将该结点插入到游标指针pos后
        // 例如: 链表1->2->3->4->5,m=2,n=4，其中待反转的部分链表为2->3->4，则从第二个结点3开始，将3插入到pos后得到1->3->2->4->5，将4插入到pos后得到1->4->3->2->1，反转操作完毕
        ListNode *orihead=pos->next;     // 反转操作前待反转部分链表的头结点
        ListNode *now=pos->next->next;   // now游标指针从待反转部分链表的第二个结点开始遍历 // 因为已经排除了m==n，所以才可以连续两次->next

        for(int i=0;i<n-m;i++)
        {
            orihead->next=now->next;
            now->next=pos->next;
            pos->next=now;

            // 更新now指针
            now=orihead->next;
        }

        // 5. 删除头部哑结点
        ListNode *temp=head;
        head=head->next;
        delete temp;

        return head;
    }
};